### PR TITLE
Adds `Progressing` status to `ExtensionsReady` seed condition 

### PR DESCRIPTION
**Pull request by plkokanov**
**How to categorize this PR?**

/area ops-productivity
/kind enhancement

**What this PR does / why we need it**:
Adds a `Progressing` status to the `ExtensionsReady` seed condition. This means that the when one of the `Valid`, `Healthy` or `Installed` conditions of any `ControllerInstallation` for a seed cluster becomes `False`, the `ExtensionsReady` condition will first be set to `Progressing`, if a configured condition threshold has not expired. Once the threshold expires, the condition will be set to `False`

Additionally, the PR moves the seed extensions check controller which is responsible for maintaining the `ExtensionsReady` condition to the gardener controller manager, as was suggested [here].(https://github.com/gardener/gardener/blob/fd3045ec72d5e11155a08c0ab56dcbe9b913f45e/pkg/gardenlet/controller/seed/seed.go#L103)

**Which issue(s) this PR fixes**:
Part of #4722

**Special notes for your reviewer**:

**Release note**:

```other operator
The `ExtensionsReady` condition for `Seed`s will first be set to `Progressing` instead of being directly set to `False` when a `ExtensionsReady` condition threshold is specified in the `controllers.seedExtensionsCheck.conditionThresholds` configuration for the gardener controller manager and that threshold has not expired yet. 
```

### CODE REVIEW

File: pkg/controllermanager/controller/seed/seed_extension_check_control.go [hunk 1]
```diff
@@ -39,8 +40,16 @@ func (c *Controller) controllerInstallationOfSeedAdd(obj interface{}) {
 	c.seedExtensionCheckQueue.Add(controllerInstallation.Spec.SeedRef.Name)
 }
 
-func (c *Controller) controllerInstallationOfSeedUpdate(_, newObj interface{}) {
-	c.controllerInstallationOfSeedAdd(newObj)
+func (c *Controller) controllerInstallationOfSeedUpdate(oldObj, newObj interface{}) {
+	oldControllerInstalaltion, ok1 := oldObj.(*gardencorev1beta1.ControllerInstallation)
+	newControllerInstalaltion, ok2 := newObj.(*gardencorev1beta1.ControllerInstallation)
```
Comment (rfranzke) [inline pkg/controllermanager/controller/seed/seed_extension_check_control.go]: typos

File: pkg/controllermanager/controller/seed/seed_extension_check_control.go [hunk 2]
```diff
@@ -39,8 +40,16 @@ func (c *Controller) controllerInstallationOfSeedAdd(obj interface{}) {
...
+	if !ok1 || !ok2 {
+		return
+	}
+
+	if !apiequality.Semantic.DeepEqual(oldControllerInstalaltion.Status.Conditions, newControllerInstalaltion.Status.Conditions) {
```
Comment (rfranzke) [inline pkg/controllermanager/controller/seed/seed_extension_check_control.go]: Should we make this check more sophisticated by only enqueuing when the `Status` of the conditions changed? Right now, even if some heart beat time or whatever time stamp will be changed we would also enqueue.

File: pkg/controllermanager/controller/seed/seed.go [hunk 3]
```diff
@@ -86,20 +89,26 @@ func NewSeedController(
...
+		extensionCheckReconciler: NewExtensionCheckReconciler(gardenClient, metav1.Now),
 
-		secretsQueue:          workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Seed Secrets"),
-		seedBackupBucketQueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Backup Bucket"),
-		seedLifecycleQueue:    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Seed Lifecycle"),
-		workerCh:              make(chan int),
+		secretsQueue:            workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Seed Secrets"),
+		seedBackupBucketQueue:   workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Backup Bucket"),
+		seedLifecycleQueue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Seed Lifecycle"),
+		seedExtensionCheckQueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "seed-extension-check"),
```
Comment (rfranzke) [inline pkg/controllermanager/controller/seed/seed.go]: Nit: For symmetry reasons:

```suggestion
		seedExtensionCheckQueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Extension Check"),
```

File: pkg/controllermanager/apis/config/types.go [hunk 4]
```diff
@@ -212,6 +214,19 @@ type SeedControllerConfiguration struct {
...
 }
 
+// SeedExtensionsCheckControllerConfiguration defines the configuration of the SeedExtensionsCheck
+// controller.
+type SeedExtensionsCheckControllerConfiguration struct {
+	// ConcurrentSyncs is the number of workers used for the controller to work on
+	// events.
+	ConcurrentSyncs *int
+	// SyncPeriod is the duration how often the existing resources are reconciled (how
+	// often the health check of Seed Extensions is performed.
```
Comment (rfranzke) [inline pkg/controllermanager/apis/config/types.go]:
```suggestion
	// often the health check of Seed Extensions is performed).
```

File: pkg/controllermanager/apis/config/v1alpha1/types.go [hunk 5]
```diff
@@ -252,6 +255,22 @@ type SeedControllerConfiguration struct {
...
 
+// SeedExtensionsCheckControllerConfiguration defines the configuration of the SeedExtensionsCheck
+// controller.
+type SeedExtensionsCheckControllerConfiguration struct {
+	// ConcurrentSyncs is the number of workers used for the controller to work on
+	// events.
+	// +optional
+	ConcurrentSyncs *int `json:"concurrentSyncs,omitempty"`
+	// SyncPeriod is the duration how often the existing resources are reconciled (how
+	// often the health check of Seed Extensions is performed.
```
Comment (rfranzke) [inline pkg/controllermanager/apis/config/v1alpha1/types.go]:
```suggestion
	// often the health check of Seed Extensions is performed).
```

File: pkg/controllermanager/apis/config/v1alpha1/defaults_test.go [hunk 6]
```diff
@@ -87,6 +87,8 @@ var _ = Describe("Defaults", func() {
 			Expect(obj.Controllers.Seed.MonitorPeriod).To(PointTo(Equal(metav1.Duration{Duration: 40 * time.Second})))
 			Expect(obj.Controllers.Seed.ShootMonitorPeriod).To(PointTo(Equal(metav1.Duration{Duration: 200 * time.Second})))
 
+			Expect(obj.Controllers.Seed).NotTo(BeNil())
```
Comment (rfranzke) [inline pkg/controllermanager/apis/config/v1alpha1/defaults_test.go]: Shouldn't it be like this?

```suggestion
			Expect(obj.Controllers.SeedExtensionsCheck).NotTo(BeNil())
```

File: pkg/controllermanager/controller/seed/seed_extension_check_control.go [hunk 7]
```diff
@@ -147,47 +159,62 @@ func (r *extensionCheckReconciler) reconcile(ctx context.Context, seed *gardenco
...
+	out := make(map[gardencorev1beta1.ConditionType]time.Duration)
+	for _, threshold := range r.config.ConditionThresholds {
+		out[gardencorev1beta1.ConditionType(threshold.Type)] = threshold.Duration.Duration
+	}
+	return out
+}
+
+// FailedCondition returns a progressing or false condition depending on the progressing threshold.
+func (r *extensionCheckReconciler) failedCondition(
+	conditionThresholds map[gardencorev1beta1.ConditionType]time.Duration,
```
Comment (rfranzke) [inline pkg/controllermanager/controller/seed/seed_extension_check_control.go]: Can you rather pass only the threshold duration for this specific condition instead of the full map? Then you can replace statements like this

```golang
		if _, ok := conditionThresholds[condition.Type]; !ok {
```

with

```golang
		if conditionThreshold == 0 {
```

File: pkg/controllermanager/controller/seed/seed_extension_check_control_test.go [hunk 8]
```diff
@@ -79,7 +82,8 @@ var _ = Describe("ExtensionCheckReconciler", func() {
 	AfterEach(func() {
 		if err := c.Get(ctx, request.NamespacedName, seed); !apierrors.IsNotFound(err) {
 			Expect(err).NotTo(HaveOccurred())
-			Expect(seed.Status.Conditions).To(ConsistOf(expectedCondition))
+			Expect(seed.Status.Conditions).NotTo(BeZero())
```
Comment (rfranzke) [inline pkg/controllermanager/controller/seed/seed_extension_check_control_test.go]:
```suggestion
			Expect(seed.Status.Conditions).NotTo(BeEmpty())
```

makes more sense / can be read easier

File: pkg/controllermanager/controller/seed/seed_extension_check_control_test.go [hunk 9]
```diff
@@ -79,7 +82,8 @@ var _ = Describe("ExtensionCheckReconciler", func() {
...
+			Expect(seed.Status.Conditions[0]).To(matchExpectedCondition)
```
Comment (rfranzke) [inline pkg/controllermanager/controller/seed/seed_extension_check_control_test.go]: Use `ConsistOf` instead like it was before? Or why did you change it?

File: pkg/controllermanager/controller/seed/seed_extension_check_control_test.go [hunk 10]
```diff
@@ -241,7 +329,61 @@ var _ = Describe("ExtensionCheckReconciler", func() {
...
+					Message:            `Some extensions are not valid: map[foo-2:]`,
+					LastTransitionTime: metav1.NewTime(fakeClock.Now().Add(-90 * time.Second)),
+					LastUpdateTime:     metav1.NewTime(fakeClock.Now().Add(-90 * time.Second)),
+				}
+				seed.Status.Conditions = []gardencorev1beta1.Condition{existingCondition}
+				Expect(c.Status().Update(ctx, seed)).To(Succeed())
+
+				matchExpectedCondition = beConditionWithStatusReasonAndMessage(gardencorev1beta1.ConditionFalse, "NotAllExtensionsHealthy", `Some extensions are not healthy: map[foo-2:]`)
+				Expect(reconciler.Reconcile(ctx, request)).To(Equal(reconcile.Result{RequeueAfter: syncPeriodDuration}))
+			})
```
Comment (rfranzke) [inline pkg/controllermanager/controller/seed/seed_extension_check_control_test.go]: Can this be made reusable by putting it into a function? This whole blocks looks very similar to what you were doing above?

File: pkg/controllermanager/controller/seed/seed_extension_check_control_test.go [hunk 11]
```diff
@@ -241,7 +329,61 @@ var _ = Describe("ExtensionCheckReconciler", func() {
...
 		})
 	})
 })
+
+func beConditionWithStatusReasonAndMessage(status gardencorev1beta1.ConditionStatus, reason, message string) types.GomegaMatcher {
```
Comment (rfranzke) [inline pkg/controllermanager/controller/seed/seed_extension_check_control_test.go]:
```suggestion
func matchConditionWithStatusReasonAndMessage(status gardencorev1beta1.ConditionStatus, reason, message string) types.GomegaMatcher {
```

File: test/integration/controllermanager/seed/extensionscheck/extensionscheck_test.go [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+	"github.com/gardener/gardener/pkg/controllermanager/apis/config"
+	"github.com/gardener/gardener/pkg/controllermanager/controller/seed"
+	"github.com/gardener/gardener/pkg/controllerutils/mapper"
+	"github.com/gardener/gardener/pkg/utils/test"
+)
+
+const (
+	conditionThreshold = 2 * time.Second
```
Comment (rfranzke) [inline test/integration/controllermanager/seed/extensionscheck/extensionscheck_test.go]: Can this be reduced to some `ms` as well? 👀 Or why did you chose `2s`?

File: test/integration/controllermanager/seed/extensionscheck/extensionscheck_test.go [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+						Services: pointer.String("100.72.0.0/13"),
+					},
+				},
+				DNS: gardencorev1beta1.SeedDNS{
+					IngressDomain: pointer.String("someingress.example.com"),
+				},
+			},
+		}
+		Expect(testClient.Create(ctx, seed)).To(Succeed())
+		By(fmt.Sprintf("seed name is %s", seed.Name))
```
Comment (rfranzke) [inline test/integration/controllermanager/seed/extensionscheck/extensionscheck_test.go]:
```suggestion
		By("Create Seed " + seed.Name)
```

File: test/integration/controllermanager/seed/extensionscheck/extensionscheck_test.go [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+		DeferCleanup(func() {
+			By("Delete Seed")
+			Expect(client.IgnoreNotFound(testClient.Delete(ctx, seed))).To(Succeed())
+		})
+
+		By("creating ControllerInstallations")
```
Comment (rfranzke) [inline test/integration/controllermanager/seed/extensionscheck/extensionscheck_test.go]:
```suggestion
		By("Create ControllerInstallations")
```

File: test/integration/controllermanager/seed/extensionscheck/extensionscheck_test.go [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+		ci2 = ci1.DeepCopy()
+		ci2.SetName("foo-2")
+
+		//This is required so that the ExtensionsReady condition is created with appropriate lastUpdateTimestamp and lastTransitionTimestamp.
+		cleanUpTestVars := test.WithVars(
+			&gardencorev1beta1helper.Now, func() metav1.Time { return metav1.Time{Time: fakeClock.Now()} },
+		)
+		DeferCleanup(func() {
+			cleanUpTestVars()
+		})
```
Comment (rfranzke) [inline test/integration/controllermanager/seed/extensionscheck/extensionscheck_test.go]:
```suggestion
		// This is required so that the ExtensionsReady condition is created with appropriate lastUpdateTimestamp and lastTransitionTimestamp.
		DeferCleanup(test.WithVars(
			&gardencorev1beta1helper.Now, func() metav1.Time { return metav1.Time{Time: fakeClock.Now()} },
		))
```

File: test/integration/controllermanager/seed/extensionscheck/extensionscheck_test.go [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+}
+
+func mapControllerInstallationToSeed(_ context.Context, _ logr.Logger, _ client.Reader, obj client.Object) []reconcile.Request {
+	controllerInstallation := obj.(*gardencorev1beta1.ControllerInstallation)
+	return []reconcile.Request{{
+		NamespacedName: types.NamespacedName{
+			Name: controllerInstallation.Spec.SeedRef.Name,
+		},
+	}}
+}
```
Comment (rfranzke) [inline test/integration/controllermanager/seed/extensionscheck/extensionscheck_test.go]: This rather belongs to the `suite_test.go` file, right?

File: test/integration/controllermanager/seed/extensionscheck/extensionscheck_test.go [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+		By(fmt.Sprintf("seed name is %s", seed.Name))
+		DeferCleanup(func() {
+			By("Delete Seed")
+			Expect(client.IgnoreNotFound(testClient.Delete(ctx, seed))).To(Succeed())
+		})
+
+		By("creating ControllerInstallations")
+		ci1 = &gardencorev1beta1.ControllerInstallation{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "foo-1",
```
Comment (rfranzke) [inline test/integration/controllermanager/seed/extensionscheck/extensionscheck_test.go]: Isn't it safer/less flake-prone if we used `GenerateName` here as well instead of the static names?

(same below for `ci2`)

**Review by rfranzke — CHANGES_REQUESTED** [review]
With @shafeeqes we checked the PR and have a few comments.
Generally, it looks quite good, thanks for also adding the integration test :)

File: test/integration/controllermanager/seed/extensionscheck/extensionscheck_test.go (hunk 12 above)
Comment (plkokanov) [reply test/integration/controllermanager/seed/extensionscheck/extensionscheck_test.go]: The minimum value that `r.clock.Now().UTC().Sub(condition.LastTransitionTime.Time.UTC())` returns seems to be 1s. I didn't pick 2 seconds for any specific reason. Setting it to 1 would also work.

File: test/integration/controllermanager/seed/extensionscheck/extensionscheck_test.go (hunk 13 above)
Comment (plkokanov) [reply test/integration/controllermanager/seed/extensionscheck/extensionscheck_test.go]: This was a left over. I will replace it with a log.Info() though as I see is done for other tests.

File: pkg/controllermanager/controller/seed/seed_extension_check_control.go (hunk 2 above)
Comment (plkokanov) [reply pkg/controllermanager/controller/seed/seed_extension_check_control.go]: Yep, but I would also include changes to the `Reason` and `Message` so that we get better error reporting and always reflect the actual error in the `SeedExtensions` condition
